File: 0050. Pow(x, n)/Solution.java
//better iterative solution  

class Solution {
    public double myPow(double x, int n) {
        double res = 1;
        long N = n;
        if(n<0){
            x = 1/x;
            N = -N;
        }

        while(N>0){
            if(N%2==1){
                res = res*x;
                N--;
            }
            else {
                x = x*x;
                N = N/2;
            }
        }
        return res;
    }
}

//bruteforce may cause TLE

// class Solution {
//     public double myPow(double x, int n) {
//         int sign = 1;
//         if(n<0)sign = -1;

//         long n1 = Math.abs(n);
//         int i = 0;
//         double ans = 1;
//         while(i<n1){
//             ans*=x;
//             i++;
//         }

//         return (sign==-1)?(1/ans):ans;
//     }
// }

File: 0050. Pow(x, n)/Solution.py
#better iterative solution  

class Solution(object):
    def myPow(self, x, n):
        """
        :type x: float
        :type n: int
        :rtype: float
        """

        N = n
        res = 1

        if(N<0):
            N = -N
            x = 1/x

        while N>0:
            if(N%2==1):
                res = res*x
                N-=1

            else:
                x = x*x
                N=N/2
        
        return res
        
#using recursion

# class Solution(object):
#     def myPow(self, x, n):
#         """
#         :type x: float
#         :type n: int
#         :rtype: float
#         """

#         N = n 
#         if N<0:
#             N = -N
#             x = 1/x

#         return self.power(x,N)



#     def power(self,x,n):
#         if n==0:
#             return 1

#         half = self.power(x,n//2)

#         if n%2:
#             return x * half * half
#         else:
#             return half*half

        

        

        

# bruteforce may cause TLE

# class Solution(object):
#     def myPow(self, x, n):
#         """
#         :type x: float
#         :type n: int
#         :rtype: float
#         """
#         sign = -1 if (n<0) else 1

#         i = 0
#         N = abs(n)
#         res = 1

#         while i<N:
#             res*=x
#             i+=1

#         return (1/res) if (sign<0) else res